parser/src/aspects/match.rs

```diff
@@ -0,0 +1,526 @@
+use lexer::token::TokenType;
+use lexer::token::TokenType::{At, Bar, CurlyLeftBracket, CurlyRightBracket, FatArrow, Identifier, If};
+
+use crate::aspects::literal::LiteralAspect;
+use crate::ast::Expr;
+use crate::ast::r#match::{Match, MatchArm, MatchPattern};
+use crate::ast::r#match::MatchPattern::{Literal, Template, VarRef, Wildcard};
+use crate::moves::{aerated, any, blanks, eox, MoveOperations, not, of_type, of_types, repeat};
+use crate::parser::{Parser, ParseResult};
+
+/// A Parser Aspect for match expression-statement and value
+pub trait MatchAspect<'a> {
+    ///parse a match statement, input parser determines how to parse each arm body
+    fn parse_match<P>(&mut self, parse_arm: P) -> ParseResult<Match<'a>>
+        where P: FnMut(&mut Self) -> ParseResult<Expr<'a>> + Clone;
+}
+
+impl<'a> MatchAspect<'a> for Parser<'a> {
+    fn parse_match<P>(&mut self, parse_arm: P) -> ParseResult<Match<'a>>
+        where P: FnMut(&mut Self) -> ParseResult<Expr<'a>> + Clone {
+        self.cursor.force(
+            of_type(TokenType::Match).and_then(blanks()),
+            "expected 'match' keyword at start of match expression.",
+        )?;
+
+        let operand = Box::new(self.expression_statement()?);
+
+        let arms = self.parse_match_arms(parse_arm)?;
+
+        Ok(Match {
+            operand,
+            arms
+        })
+    }
+}
+
+impl<'a> Parser<'a> {
+    fn parse_match_arms<P>(&mut self, parse_arm: P) -> ParseResult<Vec<MatchArm<'a>>>
+        where P: FnMut(&mut Self) -> ParseResult<Expr<'a>> + Clone {
+        self.cursor.force(blanks().then(of_type(CurlyLeftBracket)), "expected '{'")?;
+
+        let mut arms: Vec<MatchArm<'a>> = Vec::new();
+
+        while self.cursor.advance(blanks().then(of_type(CurlyRightBracket))).is_none() {
+            let arm = self.parse_match_arm(parse_arm.clone())?;
+            arms.push(arm);
+        }
```

Comment:
```suggestion
        while self.cursor.lookahead(blanks().then(eod())).is_none() {
            let arm = self.parse_match_arm(parse_arm.clone())?;
            arms.push(arm);
        }
        self.cursor.force(blanks().then(of_type(CurlyRightBracket)), "expected '}'")?;
```
The parser should stop at each closing punctuation to report unpaired delimiters (see #33).